mod_app/utils/mixins.py

```diff
@@ -0,0 +1,23 @@
+from django.utils.html import format_html
+
+
+class PreviewMixin:
+    def preview(self, obj):
+        if hasattr(obj, "file") and obj.file:
+            return self.file_preview(obj)
+        elif hasattr(obj, "url") and obj.url:
+            return self.link_preview(obj)
+        else:
+            return "-"
+
+    def file_preview(self, obj):
+        return format_html(
+            '<img src="{}" style="max-width: 15rem;" />',
+            obj.file.url,
+        )
+
+    def link_preview(self, obj):
+        return format_html(
+            '<img src="{}" style="max-width: 15rem;" />',
+            obj.url,
+        )
```

Comment:
If you put the `__init__` method in the mixin, then add it to the link classes before TabularInline, you can save yourself repeating some code:
```suggestion
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.readonly_fields += ("preview",)

    def preview(self, obj):
        if src_url := self.get_src_url(obj):
            return format_html(
                '<img src="{}" style="max-width: 15rem;" />',
                src_url,
            )
        else:
            return "-"

    def get_src_url(self, obj):
        if hasattr(obj, "file") and obj.file:
            return obj.file.url
        elif hasattr(obj, "url") and obj.url:
            return obj.url
        else:
            return None
```